pallets/xcmp-handler/src/lib.rs

```diff
@@ -168,24 +177,139 @@ pub mod pallet {
 	}
 
 	impl<T: Config> Pallet<T> {
-		pub fn calculate_xcm_fee(
+		/// Get the xcm fee and weight for a transact xcm for a given chain/currency pair.
+		pub fn calculate_xcm_fee_and_weight(
 			para_id: ParachainId,
 			currency_id: T::CurrencyId,
 			transact_encoded_call_weight: u64,
-		) -> Result<u128, DispatchError> {
+		) -> Result<(u128, u64), DispatchError> {
 			let xcm_data = XcmChainCurrencyData::<T>::get(para_id, currency_id)
 				.ok_or(Error::<T>::CurrencyChainComboNotFound)?;
-			let xcm_weight = xcm_data
+			let weight = xcm_data
 				.instruction_weight
 				.checked_add(transact_encoded_call_weight)
 				.ok_or(Error::<T>::WeightOverflow)?;
-			let fee_or_err = xcm_data
+			let fee = xcm_data
 				.fee_per_second
-				.checked_mul(xcm_weight as u128)
-				.ok_or(Error::<T>::FeeOverflow.into())
-				.map(|raw_fee| raw_fee / (WEIGHT_PER_SECOND as u128));
+				.checked_mul(weight as u128)
+				.ok_or(Error::<T>::FeeOverflow)
+				.map(|raw_fee| raw_fee / (WEIGHT_PER_SECOND as u128))?;
+
+			Ok((fee, weight))
+		}
+
+		/// Get the instructions for a transact xcm.
+		/// Currently we only support instructions if the currency is the local chain's.
+		///
+		/// Returns two instructions sets.
+		/// The first is to execute locally.
+		/// The second is to executre on the target chain.
+		pub fn get_instruction_set(
+			para_id: ParachainId,
+			currency_id: T::CurrencyId,
+			caller: T::AccountId,
+			transact_encoded_call: Vec<u8>,
+			transact_encoded_call_weight: u64,
+		) -> Result<(xcm::latest::Xcm<T::Call>, xcm::latest::Xcm<()>), DispatchError> {
+			if currency_id != T::GetNativeCurrencyId::get() {
+				Err(Error::<T>::CurrencyChainComboNotSupported)?
+			}
+
+			let (fee, weight) = Self::calculate_xcm_fee_and_weight(
+				para_id,
+				currency_id,
+				transact_encoded_call_weight,
+			)?;
+
+			let instructions = Self::get_local_currency_instructions(
+				para_id,
+				caller,
+				transact_encoded_call,
+				transact_encoded_call_weight,
+				weight,
+				fee,
+			);
+
+			Ok(instructions)
+		}
+
+		/// Construct the instructions for a transact xcm with the chain's local currency.
+		///
+		/// Local instructions
+		/// 	- WithdrawAsset
+		/// 	- DepositAsset
+		///
+		/// Target instructions
+		/// 	- ReserveAssetDeposited
+		/// 	- BuyExecution
+		/// 	- DescendOrigin
+		/// 	- Transact
+		/// 	- RefundSurplus
+		/// 	- DepositAsset
+		pub fn get_local_currency_instructions(
+			para_id: ParachainId,
+			caller: T::AccountId,
+			transact_encoded_call: Vec<u8>,
+			transact_encoded_call_weight: u64,
+			xcm_weight: u64,
+			fee: u128,
+		) -> (xcm::latest::Xcm<T::Call>, xcm::latest::Xcm<()>) {
+			// XCM for foreign chain
+			let local_asset_on_foreign = MultiAsset {
```

Comment:
i know it's less readable, but in light of your discovery on the functional programming, would it be better (i.e. faster) to just write all of the instructions in a single large array, rather than breaking out in to more readable varaibles?